### ISSUE TITLE
Color problems

### ISSUE THREAD

**Issue by TomasPuverle** [LLNL/spack#2441]
I just spent some time stepping through the code trying to debug something which turned out to be a phantom issue.

I use a "Dark Solarized" color scheme in my terminals, and when running 

    spack find --long

I could only see the packages but none of the hashes.  As you've probably guessed by now, the color of the output was so similar to the terminal background that it was not visible.

I can imagine that colors are not a priority right now, so I'm filing the issue more as a reminder, but it would be nice if in the future, at least the color map in tty/color.py could be adjusted based on the terminal, or perhaps configurable.

In the meantime, an additional solution might be to just have a --nocolor option.  It appears from the code that this would, in fact, be quite simple to do.  In some places, the "color" keyword args seems hard-coded, but I think this could just be handled in tty/color.py.  As a second benefit, it would allow to dump the output to a file; at the moment, it seems that irrespective of whether the output is a tty or not, the output is colorized (my guess is that this is because of the desire to use |less -R).

Thanks,

Tom

**Comment by TomasPuverle** [comment]
I'd like to take a stab at fixing this; looking at util/tty/color.py, there is a global attribute called "_force_color".  Unless I'm missing something, this should be a trivial change.

The question I have is whether --nocolor on the command line should take precedence over "_force_color".   Looking through the usage of this, it seems to me the answer should be "yes", as _force_color does not appear to be a "user controllable" parameter.  The only other references I can see come from OutputRedirection, which appears to be a logging context manager.

Thoughts?

**Comment by tgamblin** [comment]
@TomasPuverle: I don't have time to do this but would certainly welcome a PR.

Ideally we'd have an option on the `spack` command like many of the various GNU command line utilities support, e.g. GNU `ls` has this:

```
        --color[=WHEN]         colorize the output; WHEN can be 'always' (default
                               if omitted), 'auto', or 'never'; more info below
```

If you want to implement `--color=never` and leave the other options for others, that would be awesome.  I suspect that `--color=always` wouldn't be hard if you've already done `never`.  Spack's default is supposed to be something like `auto` (i.e. only print color if the output stream is a TTY) but I have not verified that and there aren't tests for it.  We could handle verifying that after your proposed PR.

If you are feeling ambitious, you might also consider adding a `color` option to `config.yaml`... something like:

```yaml
    color: never
```

This would control the default for `--color` while still allowing it to be overridden on the command line.  We would probably add `color: auto` to `etc/defaults/config.yaml`, which [can be overridden](https://spack.readthedocs.io/en/latest/configuration.html#configuration-scopes) in higher-precedence `config.yaml` files.

The place you'd need to touch for this one would be [`lib/spack/spack/__init__.py:112`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/__init__.py#L112), where we currently pull in settings from `config.yaml`.  I think you could set the appropriate things in `llnl.util.tty.color` from there.

Does that sound reasonable?

**Comment by TomasPuverle** [comment]
Sounds good, I'll have a go.
Thanks for your input.

**Comment by adamjstewart** [comment]
If you want to take it a step further, it would be neat to have a `colors.yaml` file where you can set each color yourself, so if you have a black background, you can set things like hashes to appear in white. But I agree that being able to turn off colors is most useful.

**Comment by tgamblin** [comment]
I was trying not to pile on, but..

1. Or you could add a `colors:` section in `config.yaml` (Below `color: auto`). There aren't that many categories -- spack has colors for the tty message types (msg, debug, warn, error), and for parts of specs (hash, version, compiler/compiler version, variants, arch).  Adding these as special format keys in try.color would probably be the way to go.
1. making tty.color handle 256color terminals would be pretty awesome.

**Comment by TomasPuverle** [comment]
Hi Adam,

Thanks for the suggestion.  I was actually planning to change Todd's proposal a bit and have a section called

color:
       when: <always/never/...>
       # Possible for future expansion to override colors

and code around that.

Further to your proposal, I was looking through the code and there are a number of places where "color" is being passed around inside of spack's lib, and more importantly, there is a predefined set of colors that are hard-coded into certain strings.  Those would have to be abstracted first, and I assume that would require some buy-in from the community on the best/most agreeable way forward.  As a final point, I am primarily a back-end developer that spends most of his time in C++, so designing UI elements is probably not my forte.  :)

Cheers,

Tom

**Comment by TomasPuverle** [comment]
Hi Todd,

Please see my message to Adam.  I picked this color issue because it's fairly small, self-contained (and I filed it), in order to get my feet wet with the community.  It was not my goal to completely refactor the color formatting.
Hope that makes sense,

Tom

**Comment by tgamblin** [comment]
@TomasPuverle: definitely!  I like the possible expansion idea (see my update). As well as the refactoring of how `color` is passed around.

I think here are some tricky parts in the code w.r.t. the color parameter. `spack find` colorizes some parts of the output separately before columnizing and then printing directly to the terminal, and finding the length of color strings may be tricky if `tty.color` is the only package that decides whether to colorize.  `colify` handles ansi colors, but not spack's color syntax (I think this is how it should probably work to avoid complicating colify).  Stuff like this is why I think `auto` is harder.

**Comment by TomasPuverle** [comment]
Do you think we need a separate setting for stderr and stdout?

Also, is Github the best place to post "design" questions like these, or is there a better forum?  (IRC etc?)

Thanks,

Tom

**Comment by adamjstewart** [comment]
GitHub is the best place. We also have a Google Group, but you're much less likely to get a response.

**Comment by junghans** [comment]
I will leave this here:
```bash
spack find -ld xsdk | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g"
```

I used this is the following case:
```bash
cmake -DCMAKE_PREFIX_PATH=$(spack find -ld xsdk | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | awk '/^[a-z]/{print $1}' | xargs -n1 spack find -p / | awk '/^[^-]/{print $2";"}' | tr -d '\n') ..
```

**Comment by tgamblin** [comment]
@junghans: is your issue fixed by #5084?

**Comment by junghans** [comment]
Yes!
